ttnn/cpp/ttnn/types.hpp

```diff
@@ -59,76 +59,72 @@ struct CoreGrid {
+                const TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED,
+                std::optional<ShardSpecBuffer> shard_parameters = std::nullopt
+            ) {
+                // Factory Method for creating Buffer objects. Will return std::shared_ptr<Buffer>.
+                // This is the only interface for creating Buffers, to ensure that users don't use
+                // the Buffer class without shared_ptr, since this class uses enable_shared_from_this.
+                std::shared_ptr<Buffer> buffer_instance(new Buffer(device, size, page_size, buffer_type, buffer_layout, shard_parameters));
+                buffer_instance->allocate(); // Asynchronously allocate the buffer
+                return buffer_instance;
+            }
+        ~Buffer() {
+            this->deallocate(); // Asynchronously clean up device state used by this buffer
         }
-        uint32_t get_buf_id() {
-            return buf_id++;
+        uint32_t address() {
+            this->allocated_.wait(false); // Wait until buffer is allocated
+            return static_cast<tt::tt_metal::Buffer*>(this)->address();
         }
 
     private:
-    std::atomic<uint32_t> buf_id = 0;
-    std::mutex map_mutex;
-    std::unordered_map<uint32_t, uint32_t> buf_id_to_address_map = {};
-};
-
-inline buffer_address_map GLOBAL_BUFFER_ADDRESS_MAP;
-
-// This buffer class is compatible with multithreaded runtime (which lives in tt_eager)
-// It is derived from the tt_metal::Buffer class, but defines its own asynchronous allocation functions
-class Buffer : public tt::tt_metal::Buffer {
-    public:
         Buffer(Device *device, uint64_t size, uint64_t page_size, const BufferType buffer_type,
                 const TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED,
                 std::optional< ShardSpecBuffer> shard_parameters = std::nullopt
             ) : tt::tt_metal::Buffer(device, size, page_size, buffer_type, buffer_layout, shard_parameters, false) {
-                this->buffer_id = GLOBAL_BUFFER_ADDRESS_MAP.get_buf_id(); // Each buffer has a unique ID
-                this->allocate();
+                // Private constructor used by factory method. Each buffer is given a unique id.
+                this->buffer_instance_id = Buffer::buf_id++;
             }
-        ~Buffer() {
-            this->deallocate();
-        }
-    private:
-        uint32_t buffer_id = 0;
+
         void allocate() {
             TT_ASSERT(this->device());
-            this->device()->push_work([this] () mutable {
-                bool bottom_up = this->buffer_type() == BufferType::DRAM;
-                tt::tt_metal::detail::AllocateBuffer(this, bottom_up);
+            this->device()->push_work([shared_buffer_ptr = shared_from_this()] () mutable {
+                bool bottom_up = shared_buffer_ptr->buffer_type() == BufferType::DRAM;
+                tt::tt_metal::detail::AllocateBuffer(shared_buffer_ptr.get(), bottom_up);
```

Comment:
perfect case if we can have version which doesn't take buffer, just calls allocate. In this case we could make address to be shared_ptr to atomic.

Reply:
I may be missing something, but I'm not sure how simply updating/assigning the address inside the worker would help us. All attributes, except for the address, are populated in the main thread even with this implementation.

Even if we passed in a `std::shared_ptr` to the buffer address member, we won't be able to guarantee that we'd have it available when we call the destructor in the main thread.

Reply:
All attributes are very simple except Buffer as a whole.